### ISSUE TITLE
Unsupported fee-on-transfer tokens

### ISSUE THREAD

**Issue by MrToph** [code-423n4/org#3]
Many protocols don't work well with tokens that take a fee on each transfer.

- [High](https://github.com/code-423n4/2021-12-pooltogether-findings/issues/30)
- [Medium](https://github.com/code-423n4/2021-12-sublime-findings/issues/142)
- [Low](https://github.com/code-423n4/2021-12-perennial-findings/issues/36)

It's a very common issue and it'd be useful to have a consensus among the judges on the severity here.

**Comment by GalloDaSballo** [comment]
My answer depends on the context.

If a smart contract can be bricked without any requirement, by using a `feeOnTransfer` token that's one thing
If the smart contract >>could<< break in its entirety if the sponsor were to use a `feeOnTransfer` token that's another thing
If the functionality would break exclusively for a `feeOnTransfer` token and the issue can be sidestepped that's another thing
Lastly, if the sponsor explicitly states they are not going to support `feeOnTransfer` or state the specific token they're going to use (assuming it's not `feeOnTransfer`)


Happy to give my thoughts in all cases, but I don't believe there's a one size fits all here

**Comment by 0xean** [comment]
Agree on the lack of one size fits all approach here.

I usually don't think its a high severity issue if the sponsor has called FOT out of scope OR explicitly stated in the natspec comments they aren't supported.   The exceptions here is determined by the blast radius of the introduction of a FOT token.  Does that simply render one contract / exchange / pair unusable or does it have system wide impacts on availability, security of funds, etc.

**Comment by GalloDaSballo** [comment]
I was made aware that @MrToph linked 3 links

To be honest these are the types of issues where I will always take more than one day to judge, as I need to think these over.

Even in the Low severity finding which I did judge, I took multiple days to properly consider whether the DOS was severe enough to warrant a Medium Severity (breaking rewards for everyone), vs breaking reward for that specific token for that specific bounty (which was the case, hence in case of revert user can still claim any other token)

I find it extremely hard to make any statement about the other two findings as the more I look into them, the more I believe I'd have to read the whole code, see all the other findings, then think about the implications before I can make a statement

**Comment by MrToph** [comment]
> I usually don't think its a high severity issue if the sponsor has called FOT out of scope OR explicitly stated in the natspec comments they aren't supported.

If the sponsor mentions it's out of scope and I still raise the issue, I always assumed it would be judged as non-critical.
Is this the case, or do you still give it a low/mid severity because the code does not explicitly protect against it?

> My answer depends on the context.

Yes, makes sense. But maybe we can find a consensus among the most common case which I think is that some storage variable is updated with the `amount` but the contract receives `amount - fees`, causing a DOS for some users at some point when they try to withdraw. I think this is the case for all three issues mentioned above - at least from the issue descriptions, haven't checked the code.

**Comment by GalloDaSballo** [comment]
My rationale about severities when dealing with feeOnTransfer, notice that I'm extrapolating my thoughts from the examples, give me different examples and I'll give you a different reply

*High*
I think if the sponsor mentions it being out of scope, but you can still break the system without any pre-condition (token whitelisted, token being used) then can still ne a High Severity Finding.

e.g. I can transfer my token, and in doing so I brick the protocol, the deployer and the users cannot do anything about it, the mere fact that I use a feeOnTransfer magically allows me to steal all funds or to permanently brick the contract

*Medium*
However, most of the times the issues are relegated to either using the `feeOnTransfer` as the "principal" token (imagine want in a vault, or cToken for a lending protocol), in that case the deployer opted into the token, so there's always that external condition. 

A vault won't work if using a feeOnTransferToken.

I believe this can be further reduced in severity if the Sponsor made it clear they won't be using that type of token.
I believe this can be put out of scope if the Sponsor specifies the exact tokens they will be using (see Yield Context), if the token is known, the system either works or doesn't, there's no in between

*Low*
See my rationale here: https://github.com/code-423n4/2021-12-perennial-findings/issues/36
A token reward that can block claiming but also: 
- Is Optional
- Can be sidestepped just by not claiming
- Doesn't block any other aspect of the protocol except itself (it's the equivalent of trying to claim a token that has transfer bricked)
- The fact that you cant' claim that doesn't prevent you from claiming other tokens
- After being made aware of the finding, a user would have to willingly send their tokens just to effectively loose them (self-loss), and they wouldn't be able to grief anybody else (sidesteppable)


*Non-Critical*
I don't have an example for it being non-critical, I think it's either a valid finding or invalid

*Invalid*
The token for the protocol is known, and it's not feeOnTransfer, the finding is invalid because you will not have a token with those properties

**Comment by MrToph** [comment]
@GalloDaSballo your descriptions mostly make sense to me

> High: I think if the sponsor mentions it being out of scope, but you can still break the system without any pre-condition (token whitelisted, token being used) then can still be a High Severity Finding.

If I understand you correctly, you're saying that it can still be high severity if 1) the protocol does not protect against using fee-on-transfer tokens 2) using such a token has side effects beyond this specific token's functionality itself (for example, if depositing a fee-on-transfer token would also break deposits or withdrawals for other unrelated tokens). I agree with _high_ severity here.  
If it would only break withdrawals for the fee-on-transfer token, I'd tend more towards _medium_.

> Medium. ... in that case the deployer opted into the token

Okay with me. Personally, I'd probably even downgrade it to _low_ if the token choice is not up to the user and the protocol deployer has to explicitly initialize a contract with a fee-on-transfer token. My reasoning would be that fee-on-transfer tokens are rare and if they intended to use fee-on-transfer tokens they would have implemented them correctly. (But I'm aware that this argument is quite opinionated and I see how one would still want to judge _medium_ in this case.). 
I'd do non-critical if they in addition mention that they won't be deploying it with fee-on-transfer tokens. I think if deployment context like that is given, it's fine to judge it non-critical/invalid as the issue is indeed invalid in this context.

> Low. ... Is Optional

"Is Optional" means deployer has to explicitly whitelist / initialize the contract with this token? Yes, as I said above, my default would be _low_ in this case

---

so if I'd sum it up, as a first guideline:
- only deployer can set the token => `low`. could be `medium` depending on what functionality breaks. `invalid` if docs mention that they don't support non-standard tokens
- the token is a function parameter chosen by users: `medium`. could be `high` if it has side effects that break other unrelated core functionality

**Comment by GalloDaSballo** [comment]
Agree with your summary @MrToph 

I don't believe I've ever seen a `high` severity valid finding for `feeOnTransfer` as all findings are typically contingent on the sponsor setting the token, or they can easily be sidestepped.

However there may be a scenario where this token completely breaks a protocol, and as such it could happen and while I've never seen it, it is best not to rule it out yet

**Comment by MrToph** [comment]
Sure, that would be the exception to the rule. I think this holds for all kinds of issues - if a user can break a protocol, it's _high_ severity.

I'm not sure if we need to explicitly mention this case in the rulebook as it's a general rule and not specific to one issue, but it's good to overcommunicate that these are just some orientation guidelines and can be overwritten by judges in specific situations as they seem fit

**Comment by kartoonjoy** [comment]
Per the Autumn 2023 C4 Supreme Court verdicts, the Supreme Court's verdict on this issue is: 

> - Generally speaking, all non-standard ERC-20 token issues should be included in a dedicated part of the analysis writeup, not submitted as high, medium, or low risk issues. The exception is an actual threat to the functionality of the protocol or funds thereof. Judges should immediately invalidate useless issues in this category.
> - Regardless of the place of submission, out of scope issues should never be submitted. Doing so is a waste of everyone’s time.
> - Rather than treat EIP-20 as an authoritative document, we recognize that the commonly understood definition is the one stated here: https://ethereum.org/en/developers/docs/standards/tokens/erc-20/#body
> - 🟢 `Recommendation:` We recommend that C4 staff include the following list of common non-standard token types in the onboarding questionnaire, asking if the sponsor wishes to specifically exclude these tokens from the scope of the contest:
> - We recognize these mechanisms as in-scope by default, and suggest that the sponsor manually opt-out of the following list, which we expect over time will change, based on common mechanisms being used
>   - Fee on Transfer
>   - Rebasing / Elastic Supply
>   - Streaming Token
>   - Revert on Zero Value
>   - Revert on Non-Zero to Non-Zero Approval
>   - Doesn’t Revert on Failure
>   - Blacklisting
>   - Token upgradeability
>   - Over 18 decimals
>   - Pausability
>   - Flash-mintable tokens
>   - Callback tokens (ERC777)
>   - non-bool returning tokens
> 
> 
> - If a mechanism is out of scope but tokens with such mechanisms can be added permissionlessly AND these would have effects on assets unrelated to the mechanism, those would still be treated in scope.


Link to verdict: https://docs.google.com/document/d/1Y2wJVt0d2URv8Pptmo7JqNd0DuPk_qF9EPJAj3iSQiE/edit#heading=h.zbsuheanhpl5
